google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/util/AppEngineUtil.java

```diff
@@ -69,34 +74,38 @@ private AppEngineUtil() {
    *
    * <p>Artifacts either target the standard or the flexible environment. All standard artifacts are
    * added. Flexible artifacts are only added if there are no other standard artifacts associated
-   * with the same module - i.e. if a module is set up for App Engine standard (has an
-   * appengine-web.xml etc.) then no option is provided to deploy any of its artifacts to the
-   * flexible environment.
+   * with the same module.
    *
-   * @return a list of {@link AppEngineArtifactDeploymentSource}'s
+   * @return a list of {@link AppEngineArtifactDeploymentSource}
    */
   public static List<AppEngineArtifactDeploymentSource> createArtifactDeploymentSources(
       @NotNull final Project project) {
     List<AppEngineArtifactDeploymentSource> sources = Lists.newArrayList();
     AppEngineProjectService projectService = AppEngineProjectService.getInstance();
-    AppEngineAssetProvider assetProvider = AppEngineAssetProvider.getInstance();
 
     for (Module module : ModuleManager.getInstance(project).getModules()) {
-      XmlFile appEngineWebXml
-          = assetProvider.loadAppEngineStandardWebXml(project, Collections.singletonList(module));
-      final AppEngineEnvironment environment
-          = projectService.getModuleAppEngineEnvironment(appEngineWebXml);
-
-      boolean isFlexCompat = projectService.isFlexCompat(appEngineWebXml);
-      boolean isStandardModule = environment.isStandard() || isFlexCompat;
-
-      Collection<Artifact> artifacts = ArtifactUtil.getArtifactsContainingModuleOutput(module);
-      for (Artifact artifact : artifacts) {
-        if ((isStandardModule && projectService.isAppEngineStandardArtifactType(artifact))
-            || (!isFlexCompat && environment.isFlexible()
-                  && projectService.isAppEngineFlexArtifactType(artifact))) {
-          sources.add(createArtifactDeploymentSource(project, artifact, environment));
-        }
+      FacetManager facetManager = FacetManager.getInstance(module);
+      if (facetManager.getFacetByType(AppEngineStandardFacet.ID) != null
+          || facetManager.getFacetByType(AppEngineFlexibleFacetType.ID) != null) {
+        final AppEngineEnvironment environment =
+            projectService.getModuleAppEngineEnvironment(module).orElseThrow(
+                () -> new RuntimeException("No environment."));
```

Comment:
Indeed. This is what I mentioned as the issues coming from having to unwrap values now using Optionals.

I never saw it happening in testing. Other alternatives are going back to nulls or having another kind of "invalid value reference".